### PR TITLE
Update loadsave.cpp to support unicode paths, cf. #24491

### PR DESCRIPTION
**Pull request by CSBVision**
Implements unicode support as discussed in #24491

### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by CSBVision** [comment]
The unicode support in `imwrite` should be there, too. If not, we missed something. The idea is to use only resolved filenames inside the static / internal functions, and to only call these with resolved filenames. As the user might call the public functions with non-resolved functions, we thought this is the best approach to avoid double resolving.

Next, adding a test with a unicode / non-standard path requires to also enable C++ 17 by default. This requires to extend `OpenCVDetectCXXCompiler.cmake` by requesting C++ 17. We are open to extend the PR, nevertheless we don't know which side effects might occur by upgrading the whole code base on all platforms to C++ 17.

File: modules/core/include/opencv2/core/utils/filesystem.hpp [hunk 1]
```diff
@@ -11,6 +11,9 @@ namespace cv { namespace utils { namespace fs {
 CV_EXPORTS bool exists(const cv::String& path);
 CV_EXPORTS bool isDirectory(const cv::String& path);
 
+CV_EXPORTS cv::String resolvePath(const cv::String& path);
+CV_EXPORTS FILE* open(const cv::String& filename, const char* mode);
```
Comment (sturkmen72) [inline modules/core/include/opencv2/core/utils/filesystem.hpp:15]: is this function needed ?

Comment (CSBVision) [reply modules/core/include/opencv2/core/utils/filesystem.hpp:15]: It's not called yet. The basic idea was to replace `fopen` with this new function, however some decoders call `fopen` by themselves, thus we have to provide a resolved path (cf. discussion in #24491). We could remove this function. However, maybe it's worth considering to migrate the `fopen` calls inside OpenCV to this one to realize unicode support wherever needed.

**Comment by vpisarev** [comment]
@CSBVision,
on cv::imwrite() — you are right, I missed it.
On C++17 for a test — you can enclose the test body into the proper `#if` (probably, . In the upcoming OpenCV 5.0 (see 5.x branch) C++ 17 is the default standard, so when this patch gets ported to 5.x (we do 4.x=>5.x merges on a regular basis), the test will be 'activated'.

Actually, since OpenCV 4.x uses C++ 11 by default and will likely stay at C++ 11, maybe it's a better idea to retarget this PR to 5.x branch instead of 4.x. But feel free to keep 4.x as the base branch.

**Comment by CSBVision** [comment]
Alright, we just added the test to `test_read_write.cpp`. Additionally, we think it's worth keeping the patch in the 4.x branch as we think that it might be possible to upgrade to C++ 17 by defining the CMake flags accordingly (should be working if there are no incompatibilities). In any case, we need to define `/Zc:__cplusplus` for MSVC, otherwise the `if __cplusplus >= 201703L` checks don't work as MSVC does not define `__cplusplus` according to the actual used standard without this flag. We can also add this to the configuration flags right now and extend the PR (it won't harm anything, even though it's also not required for the moment). But, without adding it while upgrading to C++ 17, the unicode support won't be activated.

**Comment by CSBVision** [comment]
Do you still have any intention to accept this patch or should we simply close it?
We think it's still reasonable to prepare Unicode support and slightly improve the codec resolving, however we migrated all our code from `imread` and `imwrite` to `imdecode` and `imencode`, respectively, in the meanwhile (easily possible for Python, too, using `numpy.fromfile` and `numpy.tofile`, which in particular turned out to be significantly faster for massive loading). Still, we would like to finalize this patch.
